system-contracts/contracts/BatchAggregator.sol

```diff
@@ -0,0 +1,354 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {IBatchAggregator} from "./interfaces/IBatchAggregator.sol";
+import {ISystemContract} from "./interfaces/ISystemContract.sol";
+import {L2_L1_LOGS_TREE_DEFAULT_LEAF_HASH, L2_TO_L1_LOG_SERIALIZE_SIZE, STATE_DIFF_COMPRESSION_VERSION_NUMBER} from "./interfaces/IL1Messenger.sol";
+import {SystemLogKey, SYSTEM_CONTEXT_CONTRACT, KNOWN_CODE_STORAGE_CONTRACT, COMPRESSOR_CONTRACT, BATCH_AGGREGATOR, STATE_DIFF_ENTRY_SIZE, L2_TO_L1_LOGS_MERKLE_TREE_LEAVES, PUBDATA_CHUNK_PUBLISHER, COMPUTATIONAL_PRICE_FOR_PUBDATA} from "./Constants.sol";
+import {UnsafeBytesCalldata} from "./libraries/UnsafeBytesCalldata.sol";
+import {ICompressor, OPERATION_BITMASK, LENGTH_BITS_OFFSET, MAX_ENUMERATION_INDEX_SIZE} from "./interfaces/ICompressor.sol";
+
+uint256 constant BYTE_BITMASK = (1<<8)-1;
+uint256 constant DERIVED_KEY_SIZE = 32;
+uint256 constant LOOSE_COMPRESION = 33;
+contract BatchAggregator is IBatchAggregator, ISystemContract {
+    using UnsafeBytesCalldata for bytes;
+    bytes[] batchStorage;
+    bytes[] chainData;
+    // log data
+    mapping(uint256 => bytes[]) messageStorage;
+    mapping(uint256 => bytes[]) logStorage;
+    mapping(uint256 => bytes[]) bytecodeStorage;
+    // state diff data
+    mapping(uint256 => bytes[]) initialWrites;
+    mapping(uint256 => bytes32[]) initialWritesSlots; 
+    mapping(uint256 => mapping(uint64 => bytes)) uncompressedWrites;
+    mapping(uint256 => mapping(uint64 => bool)) slotStatus;
+    mapping(uint256 => uint64[]) accesedSlots;
+    
+    // chain data
+    mapping(uint256 => bool) chainSet;
+    uint256[] chainList;
+    function addChain(uint256 chainId) internal {
+        if (chainSet[chainId] == false) {
+            chainList.push(chainId);
+            chainSet[chainId] = true;
+        }
+    }
+  
+    function _sliceToUint256(bytes calldata _calldataSlice) internal pure returns (uint256 number) {
+        number = uint256(bytes32(_calldataSlice));
+        number >>= (256 - (_calldataSlice.length * 8));
+    }
+
+    function addInitialWrite(uint256 chainId, bytes calldata stateDiff, bytes calldata compressedStateDiff) internal{
+        bytes32 derivedKey = stateDiff.readBytes32(52);
+        initialWrites[chainId].push(stateDiff);
+        initialWritesSlots[chainId].push(derivedKey);
+    }
+    function addRepeatedWrite(uint256 chainId, bytes calldata stateDiff, bytes calldata compressedStateDiff) internal{
+        uint64 enumIndex = stateDiff.readUint64(84);
+        if (slotStatus[chainId][enumIndex]==false){
+            uncompressedWrites[chainId][enumIndex] = stateDiff;
```

Comment:
why do we store the entire state diff? not too bad, but I think storing just `initialValue`,`finalValue`,`enumIndex` will be much easier to read

Reply:
ok, done